### PR TITLE
refactor index/ into spark/ and add tests for curves

### PR DESCRIPTION
**Pull request by sandan**
So here's a summary of the changes

The index/ subproject was refactored into the spark/ subproject in the following way:
-The Zcurves in Index/ were moved into spark/io/index/zcurve. Similarly for the tests.
-spark/io/index/ now has additional packages hilbert/, zcurve/, and rowmajor/ each of which contain the SpatialKeyIndex and SpaceTimeKeyIndex implementations for those curves.
-This affected import statements and package names of various other classes such as the KeyIndexMethods in spark/io/index/ and the RasterRDDWriterProvider and RasterRDDReaderProvider in spark/io/accumulo.

-Tests for the ZSpatialKeyIndex and ZSpaceTimeKeyIndex are implemented
-Tests for the HilbertSpatialKeyIndex and HilbertSpaceTimeKeyIndex are implemented
-Tests for the RowMajorSpatialKeyIndex are implemented

### CODE REVIEW

**Comment by lossyrob** [comment]
This needs to be updated with the changes in catalog-refactor.

Also there are commits against emails that github doesn't recognize. You need to amend these commits.
http://stackoverflow.com/questions/750172/change-the-author-of-a-commit-in-git

File: README.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 [![Join the chat at https://gitter.im/geotrellis/geotrellis](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/geotrellis/geotrellis?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
 
-*GeoTrellis* is a Scala library and framework for creating processes to work with raster data.
+GeoTrellis is a Scala library and framework for creating processes to work with raster data.
```
Comment (lossyrob) [inline README.md]: Revert odd changes like this.

File: spark/src/main/scala/geotrellis/spark/SpatialKey.scala [hunk 2]
```diff
@@ -34,6 +34,7 @@ object SpatialKey {
 }
 
 /** A SpatialKey designates the spatial positioning of a layer's tile. */
+//this order is confusing...
```
Comment (lossyrob) [inline spark/src/main/scala/geotrellis/spark/SpatialKey.scala]: Take out comment.

**Comment by lossyrob** [comment]
Can you merge in teh current catalog-refactor into this PR and address the above comments? Thanks.

**Comment by sandan** [comment]
@lossyrob Sorry Rob, rebasing is a bit trickier than I thought, I'll need more time to make sure I get it right. I think I'd be more effective working on the benchmarks. I'm kind of pressed for time.

**Comment by sandan** [comment]
@lossyrob I've taken care of the commits you mentioned and some of the other things we've talked about (moving files using git mv). Could you elaborate more on how you want me to merge the geotrellis branch into this PR?

**Comment by lossyrob** [comment]
```
git fetch upstream
git merge upstream/feature/catalog-refactor
(solve any conflicts)
git push origin feature/catalog-refactor
```

If you're wondering how to solve the conflicts, it's pretty simple...it'll say which files have merge conflicts. Those files will be edited with both version of the code inside a format that shows what's the incoming change and what's the code that's conflicting. You just have to modify those files to have the incoming changes without the conflict. This might be informative: http://githowto.com/resolving_conflicts

Does that help? What specific problems are you having?

**Comment by sandan** [comment]
@lossyrob That helps, I'll work on it today.

File: spark/src/test/scala/geotrellis/spark/io/index/hilbert/HilbertSpaceTimeKeyIndexSpec.scala [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+import org.scalatest._
+import org.joda.time.DateTime
+import scala.collection.immutable.TreeSet
+import geotrellis.spark.SpaceTimeKey
+import geotrellis.spark.KeyBounds
+
+class HilbertSpaceTimeKeyIndexSpec extends FunSpec with Matchers{
+
+  val UpperBound: Int = 16
```
Comment (lossyrob) [inline spark/src/test/scala/geotrellis/spark/io/index/hilbert/HilbertSpaceTimeKeyIndexSpec.scala]: lowercase U

File: spark/src/test/scala/geotrellis/spark/io/index/hilbert/HilbertSpaceTimeKeyIndexSpec.scala [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+     
+     //check for consecutivity
+     val itr: Iterator[Long] = ts.iterator
+     var s = itr.next
+     while(itr.hasNext){
+      var t = itr.next 
+      t should be (s+1)
+      s = t
+     }
+    }
```
Comment (lossyrob) [inline spark/src/test/scala/geotrellis/spark/io/index/hilbert/HilbertSpaceTimeKeyIndexSpec.scala]:
```scala
it("Generates a Long index given a SpatialKey"){
  val rmajor = new RowMajorSpatialKeyIndex(KeyBounds[SpatialKey](SpatialKey(0,0), SpatialKey(UpperBound-1, UpperBound-1)))

  val keys = 
    for(col <- 0 until UpperBound;
         row <- 0 until UpperBound) yield {
      rmajor.toIndex(SpatialKey(col, row))
    }
  
  keys.distinct.size should be (UpperBound * UpperBound)
  keys.min should be (0)
  keys.max should be (UpperBound * UpperBound - 1)
}
```

File: spark/src/test/scala/geotrellis/spark/io/index/rowmajor/RowMajorSpatialKeyIndexSpec.scala [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+      
+       //check for consecutivity
+       val itr: Iterator[Long] = ts.iterator
+       var s = itr.next
+       while(itr.hasNext){
+        var t = itr.next
+        t should be (s+1)
+        s = t
+       }
+   }
```
Comment (lossyrob) [inline spark/src/test/scala/geotrellis/spark/io/index/rowmajor/RowMajorSpatialKeyIndexSpec.scala]:
```scala
it("Generates a Long index given a SpatialKey"){
  val rmajor = new RowMajorSpatialKeyIndex(KeyBounds[SpatialKey](SpatialKey(0,0), SpatialKey(UpperBound-1, UpperBound-1)))

  val keys = 
    for(col <- 0 until UpperBound;
         row <- 0 until UpperBound) yield {
      rmajor.toIndex(SpatialKey(col, row))
    }
  
  keys.distinct.size should be (UpperBound * UpperBound)
  keys.min should be (0)
  keys.max should be (UpperBound * UpperBound - 1)
}
```

File: spark/src/test/scala/geotrellis/spark/io/index/hilbert/HilbertSpaceTimeKeyIndexSpec.scala [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+   it("generates hand indexes you can hand check 2x2x2"){
+     val hilbert = HilbertSpaceTimeKeyIndex(SpaceTimeKey(0,0,y2k), SpaceTimeKey(2,2,y2k.plusMillis(1)),1,1) 
+     val idx = List[SpaceTimeKey]( SpaceTimeKey(0,0,y2k), SpaceTimeKey(0,1,y2k),
+                                   SpaceTimeKey(1,1,y2k), SpaceTimeKey(1,0,y2k),
+                                   SpaceTimeKey(1,0,y2k.plusMillis(1)), SpaceTimeKey(1,1,y2k.plusMillis(1)),
+                                   SpaceTimeKey(0,1,y2k.plusMillis(1)), SpaceTimeKey(0,0,y2k.plusMillis(1)))
+     
+     for{ i<-0 to 7 }
+       hilbert.toIndex(idx(i)) should be (i)
```
Comment (lossyrob) [inline spark/src/test/scala/geotrellis/spark/io/index/hilbert/HilbertSpaceTimeKeyIndexSpec.scala]:
```scala
for(i <- 0 to 7) {
  hilbert.toIndex(idx(i)) should be (i)
}
```

**Comment by sandan** [comment]
@lossyrob I've cleaned up the code and the tests are passing

**Comment by lossyrob** [comment]
Nice, it's looking good. I'll have to plan this merge with a couple other PR's. I'll let you know if there's any other updates needed.

**Comment by lossyrob** [comment]
Closing, updated in #1083

**Pull request closed without merging**